[HDR] ASSERT fires when rendering an HDR image on unaccelerated GraphicsContext

shallawa opened the pull request:
#### 428839657cf0383d37bd2a3e46fe1f8acfd5441b
<pre>
[HDR] ASSERT fires when rendering an HDR image on unaccelerated GraphicsContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=291471">https://bugs.webkit.org/show_bug.cgi?id=291471</a>
<a href="https://rdar.apple.com/149123824">rdar://149123824</a>

Reviewed by Simon Fraser.

ShareableBitmap backend should be able to create RGBA16F buffers.

Test: fast/images/hdr-unaccelerated-basic-image.html
* LayoutTests/TestExpectations:
* LayoutTests/fast/images/hdr-unaccelerated-basic-image-expected.html: Added.
* LayoutTests/fast/images/hdr-unaccelerated-basic-image.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::create):

Canonical link: <a href="https://commits.webkit.org/300715@main">https://commits.webkit.org/300715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/128f54d46c9331ec12e57281dac452f320faa431

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123594 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43309 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34005 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130361 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75737 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ef91d4b6-2250-44f0-a7d1-55e06a9abc81) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51903 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93974 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62375 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2d288430-de8c-4c54-8387-9549983c4f07) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110568 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74578 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/60622eed-b81d-4590-af43-4ef6cb49fb15) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34045 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28726 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73841 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104797 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28950 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133054 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38482 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102448 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50920 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106789 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102290 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26002 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47640 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25881 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50399 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56161 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49873 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53220 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51548 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
```diff
@@ -150,6 +150,18 @@ std::optional<DestinationColorSpace> DestinationColorSpace::asRGB() const
+        return *this;
+
+#if HAVE(CORE_GRAPHICS_EXTENDED_SRGB_COLOR_SPACE)
+    return DestinationColorSpace::ExtendedSRGB();
```
Don't we need to use `CGColorSpaceCreateExtended()` as some other PR did?

shallawa replied:
Yes we do need to use `CGColorSpaceCreateExtended()`. This change was submitted in https://commits.webkit.org/295899@main.

smfr commented on Source/WebCore/platform/graphics/ContentsFormat.cpp:
```diff
@@ -39,14 +39,18 @@ std::optional<DestinationColorSpace> contentsFormatExtendedColorSpace(ContentsFo
+#if ENABLE(PIXEL_FORMAT_RGBA16F)
     case ContentsFormat::RGBA16F:
-        return DestinationColorSpace::ExtendedRec2020();
+        return DestinationColorSpace::ExtendedSRGB();
```
I can't tell if this will clamp P3 colors.

shallawa replied:
Using ExtendedSRGB is what we agreed on using for both RGBA10 and RGBA16F. This change was submitted in  https://commits.webkit.org/296989@main.

shallawa committed:
[HDR] ASSERT fires when rendering an HDR image on unaccelerated GraphicsContext
https://bugs.webkit.org/show_bug.cgi?id=291471
rdar://149123824

Reviewed by Simon Fraser.

ShareableBitmap backend should be able to create RGBA16F buffers.

Test: fast/images/hdr-unaccelerated-basic-image.html
* LayoutTests/TestExpectations:
* LayoutTests/fast/images/hdr-unaccelerated-basic-image-expected.html: Added.
* LayoutTests/fast/images/hdr-unaccelerated-basic-image.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::create):

Canonical link: https://commits.webkit.org/300715@main
